### PR TITLE
Accessibility for tab component

### PR DESCRIPTION
**Pull request by RicardoRFaria**
We added some features to provide accessibility in tabs.

- Aria attributes to describe the component and your states for screen readers;
- Arrow navigation like:
   * Tab to change focus from tab to your container
   * Top and left to backward, right and down to forward
   * Enter to select one tab

The documentation has been updated and tests were created.
Referencing the discussion #3526

### CODE REVIEW

File: src/tabs/docs/readme.md [hunk 1]
```diff
@@ -25,11 +25,15 @@ AngularJS version of the tabs directive.
  * `active` <i class="glyphicon glyphicon-eye-open"></i>
  	_(Defaults: false)_ :
  	Whether tab is currently selected.
-
+
```
Comment (wesleycho) [inline src/tabs/docs/readme.md]: Is this change necessary?

File: src/tabs/tabs.js [hunk 2]
```diff
@@ -288,6 +294,34 @@ angular.module('ui.bootstrap.tabs', [])
...
+  function addNavigationHandler(tab) {
+    var topParent = tab.parent().parent();
+    var tabsList = $(topParent).find('ul:first');
+    //set keydown events on tabList item for navigating and selection tabs
+    $(tabsList).on('keydown', 'a',
+      function (e) {
+        switch (e.which) {
+          case 37: case 38:
+          if ($(this).parent().prev().length !== 0) {
+            $(this).parent().prev().find('>a').focus();
```
Comment (wesleycho) [inline src/tabs/tabs.js]: This will not work if the developer is not using jQuery.

**Comment by wesleycho** [comment]
I am not sure that this is the best way to handle this.

Also the history needs to be fixed (there should not be a merge commit).

File: src/tabs/tabs.js (hunk 2 above)
Comment (RicardoRFaria) [inline src/tabs/tabs.js]: Sorry, I forgot that ui-bootstrap doesn't require Jquery.
I will fix this.

**Comment by RicardoRFaria** [comment]
Hi wesley, I fixed the history and migrate key navigation to pure javascript.
What you think it would be the best way to add accessibility for this component?

**Comment by RicardoRFaria** [comment]
I made a merge to update this branch and squashed the commits.

File: src/tabs/tabs.js [hunk 3]
```diff
@@ -192,6 +193,7 @@ angular.module('ui.bootstrap.tabs', [])
     scope: {
       active: '=?',
       heading: '@',
+      id : '@',
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Should be `@?`

File: src/tabs/tabs.js [hunk 4]
```diff
@@ -230,11 +232,22 @@ angular.module('ui.bootstrap.tabs', [])
         }
       };
 
+      scope.select = function() {
+        if ( !scope.disabled ) {
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Remove extra whitespace in parentheses

File: src/tabs/tabs.js [hunk 5]
```diff
@@ -230,11 +232,22 @@ angular.module('ui.bootstrap.tabs', [])
...
+          scope.active = true;
+        }
+      };
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Add a new line after this line.

File: src/tabs/tabs.js [hunk 6]
```diff
@@ -230,11 +232,22 @@ angular.module('ui.bootstrap.tabs', [])
...
       tabsetCtrl.addTab(scope);
       scope.$on('$destroy', function() {
         tabsetCtrl.removeTab(scope);
-      });
+      });      if ( attrs.id ) {        elm.removeAttr('id');
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Fix code style.

File: src/tabs/tabs.js [hunk 7]
```diff
@@ -230,11 +232,22 @@ angular.module('ui.bootstrap.tabs', [])
...
+      }
 
+      scope.select = function() {
+        if ( !scope.disabled ) {
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Fix code style

File: src/tabs/tabs.js [hunk 8]
```diff
@@ -230,11 +232,22 @@ angular.module('ui.bootstrap.tabs', [])
...
+          scope.active = true;
+        }
+      };
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Add new line after this line.

File: src/tabs/tabs.js [hunk 9]
```diff
@@ -286,6 +299,48 @@ angular.module('ui.bootstrap.tabs', [])
       node.tagName.toLowerCase() === 'data-tab-heading'
     );
   }
+
+  function addNavigationHandler(tab) {
+    var aElements = tab.parent().parent().find('ul').find('a');
```
Comment (wesleycho) [inline src/tabs/tabs.js]: This is too heavily tied to the DOM structure - this needs to change.

File: src/tabs/tabs.js [hunk 10]
```diff
@@ -286,6 +299,48 @@ angular.module('ui.bootstrap.tabs', [])
...
+    //set keydown events on tabList item for navigating and selection tabs
+    aElements.on('keyup', function (e) {
+        var aElement = e.target;
+        var tabElement = aElement.parentNode;
+        switch (e.which) {
+          case 37:case 38:
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Fix code style.

File: src/tabs/tabs.js [hunk 11]
```diff
@@ -286,6 +299,48 @@ angular.module('ui.bootstrap.tabs', [])
...
+          if (previousElementSibling(tabElement) != null) {
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Fix to `!==`

File: src/tabs/tabs.js [hunk 12]
```diff
@@ -286,6 +299,48 @@ angular.module('ui.bootstrap.tabs', [])
...
+            previousElementSibling(tabElement).querySelector('a').focus();
+          } else {
+            aElements[aElements.length - 1].focus();
+          }
+          break;
+          case 39: case 40:
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Fix code style.

File: src/tabs/tabs.js [hunk 13]
```diff
@@ -286,6 +299,48 @@ angular.module('ui.bootstrap.tabs', [])
...
+          if (nextElementSibling(tabElement) != null) {
```
Comment (wesleycho) [inline src/tabs/tabs.js]: Whole block should be indented one more level, and never do `!=` over `!==`

File: src/tabs/test/tabs.spec.js [hunk 14]
```diff
@@ -96,6 +96,72 @@ describe('tabs', function() {
       titles().eq(1).find('a').click();
       expect(scope.deselectFirst).toHaveBeenCalled();
     });
+
```
Comment (wesleycho) [inline src/tabs/test/tabs.spec.js]: Remove extra line

File: src/tabs/test/tabs.spec.js [hunk 15]
```diff
@@ -96,6 +96,72 @@ describe('tabs', function() {
...
+  });
+
+  describe('accessibility', function() {
+
```
Comment (wesleycho) [inline src/tabs/test/tabs.spec.js]: Remove extra line

File: template/tabs/tab.html [hunk 16]
```diff
@@ -1,3 +1,2 @@
-<li ng-class="{active: active, disabled: disabled}">
-  <a href ng-click="select()" tab-heading-transclude>{{heading}}</a>
-</li>
+<li ng-class="{active: active, disabled: disabled}" role="presentation">
+  <a href="#" ng-click="$event.preventDefault(); select()" role="tab" tabindex="{{active ? 0 : -1}}" aria-selected="{{!!active}}" id="{{id}}" tab-heading-transclude>{{heading}}</a></li>
```
Comment (wesleycho) [inline template/tabs/tab.html]: Remove `="#"` from `href`.

File: template/tabs/tabset.html [hunk 17]
```diff
@@ -1,10 +1,9 @@
 <div>
-  <ul class="nav nav-{{type || 'tabs'}}" ng-class="{'nav-stacked': vertical, 'nav-justified': justified}" ng-transclude></ul>
+  <ul class="nav nav-{{type || 'tabs'}}" ng-class="{'nav-stacked': vertical, 'nav-justified': justified}" role="tablist" ng-transclude></ul>
   <div class="tab-content">
-    <div class="tab-pane" 
-         ng-repeat="tab in tabs" 
-         ng-class="{active: tab.active}"
-         tab-content-transclude="tab">
+    <div class="tab-pane" ng-repeat="tab in tabs" ng-class="{active: tab.active}"
```
Comment (wesleycho) [inline template/tabs/tabset.html]: Why was the code style of the template changed? Revert to old style.

**Comment by wesleycho** [comment]
There are a lot of code style problems here - most problematic though is the DOM traversal mentioned https://github.com/angular-ui/bootstrap/pull/3908#discussion_r36631832 . That would be a breaking change since if a user is overriding the template, it enforces DOM structure on that template.

**Comment by RicardoRFaria** [comment]
Sorry wesley, automatic merge made some changes in code style that I didn't see. My fault.
About DOM traversal problem, I will need more time to think about it. I trought that if add event to each 'a' element after tab is added we would avoid ```.parent().parent()```, but I need more time to test.

**Comment by icfantv** [comment]
Closing in favor of #4772.
